builders/container-apps-internal-registry-demo/startup-script.sh

```diff
@@ -52,49 +52,124 @@ token=$(printf "%s" "$REGISTRY_AUTH_USERNAME:$REGISTRY_AUTH_PASSWORD" | base64)
 acr_access_string="Basic $token"
 export CNB_REGISTRY_AUTH='{"'$ACR_RESOURCE_NAME'":"'$acr_access_string'"}'
 
+echo "Initiating buildpack build..."
+echo "Correlation id: '$CORRELATION_ID'"
+echo 
+
+RETRY_DELAY=2
+RETRY_ATTEMPTS=5
+
+function fail_if_retry_exceeded() {
+  retries=$1
+  if [ "$retries" -ge $RETRY_ATTEMPTS ]; then
+    echo "----- Retry attempts exceeded -----"
+    exit 1
+  fi
+}
+
 # Execute the analyze phase
 echo
-echo "======================================="
 echo "===== Executing the analyze phase ====="
-echo "======================================="
-/lifecycle/analyzer \
-  -log-level debug \
-  -run-image mcr.microsoft.com/oryx/builder:stack-run-debian-bullseye-20230926.1 \
-  $APP_IMAGE
+retryCount=0
+until [ "$retryCount" -ge $RETRY_ATTEMPTS ]
+do
+  if [ "$retryCount" -ge 1 ]; then
+    echo "----- Retrying analyze phase (attempt $retryCount) -----"
+  fi
+
+  /lifecycle/analyzer \
+    -log-level debug \
+    -run-image mcr.microsoft.com/oryx/builder:stack-run-debian-bullseye-20230926.1 \
+    $APP_IMAGE \
+    && break
+
+  retryCount=$((retryCount+1))
+  sleep $RETRY_DELAY
+done
+
+fail_if_retry_exceeded $retryCount
 
 # Execute the detect phase
 echo
-echo "======================================"
 echo "===== Executing the detect phase ====="
-echo "======================================"
-/lifecycle/detector \
-  -log-level debug \
-  -app $CNB_APP_DIR
+retryCount=0
+until [ "$retryCount" -ge $RETRY_ATTEMPTS ]
+do
+  if [ "$retryCount" -ge 1 ]; then
+    echo "----- Retrying detect phase (attempt $retryCount) -----"
+  fi
+
+  /lifecycle/detector \
+    -log-level debug \
+    -app $CNB_APP_DIR \
+    && break
+
+  retryCount=$((retryCount+1))
+  sleep $RETRY_DELAY
+done
+
+fail_if_retry_exceeded $retryCount
 
 # Execute the restore phase
 echo
-echo "======================================="
 echo "===== Executing the restore phase ====="
-echo "======================================="
-/lifecycle/restorer \
-  -log-level debug \
-  -build-image mcr.microsoft.com/oryx/builder:stack-build-debian-bullseye-20230926.1
+retryCount=0
+until [ "$retryCount" -ge $RETRY_ATTEMPTS ]
+do
+  if [ "$retryCount" -ge 1 ]; then
+    echo "----- Retrying restore phase (attempt $retryCount) -----"
+  fi
+
+  /lifecycle/restorer \
+    -log-level debug \
+    -build-image mcr.microsoft.com/oryx/builder:stack-build-debian-bullseye-20230926.1 \
+    && break
+
+  retryCount=$((retryCount+1))
+  sleep $RETRY_DELAY
+done
+
+fail_if_retry_exceeded $retryCount
 
 # Execute the extend phase
 echo
-echo "======================================"
 echo "===== Executing the extend phase ====="
```

Comment:
The extend phase is when the code is being built. We cannot retry it, or else in case of build failure we would slow down the experience too much. From my experience I've not seen it failing without a good reason.

Reply:
I'll remove this retry then, good point

Reply:
done